# Maize_Rhizobiome_2022


## analysis

The code to generate the published figures is found here in the following scripts:

**Figure 1A:** ordination_analysis.R
**Figure 1B:** microbial_groups.R
**Figure 1C:** heritability_analysis.R
**Figure 3A:** gwas_analysis.R
**Figure 3BC:** gene_expression_analysis.R
**Figure 4:** abundance_vs_phenotype.R
**Figure 5:** f_comamonadaceae_example.R
**Supplementary Figure 1:** (manually drawn)
**Supplementary Figure 2A:** ordination_analysis.R
**Supplementary Figure 2B:** heritability_analysis.R
**Supplementary Figure 3:** heritability_analysis.R
**Supplementary Figure 4:** abundance_vs_phenotype.R
**Supplementary Figure 5:** (manually drawn)
**Supplementary Figure 6:** ordination_analysis.R
**Supplementary Figure 7B:** ordination_analysis.R
**Supplementary Figure 8:** microbial_groups.R
**Supplementary Figure 9ABC:** ordination_analysis.R
**Supplementary Figure 9DE:** gwas_analysis.R

## cache

intermediate results generated by the analysis scripts

## data

**abundance_vs_phenotype.rda:** BLUPs of microbe abundance and plant traits for each maize genotype and N treatment

**asv_taxonomy.rda :** taxonomic information of ASVs

**B73_RefGen_v4_gene_ids.rda :** gene IDs from maize B73 reference genome v4

**B73_RefGen_v4_gene_pos.rda :** gene positions from maize B73 reference genome v4

**blup_lowN_150_tax_groups.csv :** BLUPs of microbe abundance log(relative abundance) for 150 microbial groups in 226 maize genotypes under -N treatment

**blup_lowN_3618_asvs.csv :** BLUPs of microbe abundance log(relative abundance) for 3618 ASVs in 226 maize genotypes under -N treatment

**blup_stdN_150_tax_groups.csv :** BLUPs of microbe abundance log(relative abundance) for 150 microbial groups in 226 maize genotypes under +N treatment

**blup_stdN_3618_asvs.csv :** BLUPs of microbe abundance log(relative abundance) for 3618 ASVs in 226 maize genotypes under +N treatment

**corr_data.rda:** correlation of microbe abundance with phenotypes for 2 N treatments and 150 microbial groups

**group_data.rda:** summary statistics for all 150 microbial groups
	H2_stdN_19, H2_lowN_19, p_values_stdN, p_values_lowN

**gwas_dat_threshold7:** Gwas results (combined separate GWAS for +N and -N) showing 1089 SNPs that had a significant signal (-log10(p) > 7.187346) for any of the 150 microbial groups

**mean_gene_expression_by_tissue_type_4genotypes.rda:** Gene expression of genes near MAPLs vs other genes in 4 genotypes and 2 tissue types(growth chamber experiment)

**mean_gene_expression_by_tissue_type_kremling.rda:** Gene expression of genes near MAPLs vs other genes in all genotypes and 7 tissue types (data by Kremling et. al.)

**ps_asv.rda:** R phyloseq object, ASV table with 3626 ASVs and 3306 samples, final set of ASVs table used in this study (after filter step 8 in Supplementary Methods).

**ps_bothyears.rda:** R phyloseq object, ASV table with 3728 ASVs and 3307 samples, filtered ps_common.rda, includes all ASVs that were observed at least once in both years (after filter step 7 in Supplementary Methods). 

**ps_grp.rda:** R phyloseq object, feature table where ps_asv was agglomerated into the 150 microbial groups by adding the counts of all ASVs in each group. Table with 150 microbial groups (identified by a representative ASV id, see supplementary file 2 for taxonomy) and 3306 samples.

**ps_common.rda:** R phyloseq object, ASV table with 4632 ASVs and 3307 samples. Removed ASVs that were not observed in at least 5% (166) of all samples (after filter step 6 in Supplementary Methods). Includes phylogenetic tree generated with mafft/7.407 and fasttree/2.1

**ps_noMC.rda:** R phyloseq object, ASV table with 129767 ASVs and 3313 samples, filtered ASV table with mitochondria and chloroplast sequences removed (after filter step 5 in Supplementary Methods)

**V3_V4.csv :** gene names in B73 reference genome v3 and v4


## figures

Figures generated by the analysis scripts are stored here

## largedata

This directory is not uploaded to GitHub (too large).
Download this directory here (LINK) for the following files:

**GWAS/:** GWAS results in the form of .assoc.txt files

**rna_seq_4genotypes.csv :** FPKM reads of gene expression data (growth chamber experiment)
**rna_seq_kremling.txt :** FPKM reads of gene expression data by Kremling et.al.

**hapmap_chr10_23800k-24100k.txt:** Cutout of the  chr 10 allele information, derived from HapMap3





